Rector itself - merge isCandidate() to refactor() + add getNodeType()

TomasVotruba opened the pull request:
Closes #553

carusogabriel commented on composer.json:
```diff
@@ -45,7 +45,8 @@
             "Rector\\YamlRector\\": "packages/YamlRector/src"
-        }
+        },
+        "classmap": ["custom"]
```
Why?

TomasVotruba replied:
So composer can load `custom/MergeIsCandidateRector.php`.

I'll move it to `/examples` in the end, so we have complex code to show people.

carusogabriel commented on custom/MergeIsCandidateRector.php:
```diff
@@ -0,0 +1,317 @@
+use Rector\RectorDefinition\RectorDefinition;
+use Rector\Utils\NodeTraverser\CallableNodeTraverser;
+
+final class MergeIsCandidateRector extends AbstractRector
```
Will we drop it before merge this PR?

TomasVotruba replied:
Yea, I'll just move it to `/examples`, to show what Rector can do.

carusogabriel commented on packages/PHPUnit/src/Rector/ExceptionAnnotationRector.php:
```diff
@@ -73,30 +73,23 @@ public function test()
      */
     public function refactor(Node $classMethodNode): ?Node
     {
+        if (! $classMethodNode instanceof ClassMethod) {
```
I don't think we need it. If the refactor method was called, it means we have an instance of `ClassMethod`, like returned in `getNodeTypes`. This should be fixed in all Rectors

TomasVotruba replied:
Yes yes, working on it!

carusogabriel commented on src/Rector/AbstractRector.php:
```diff
@@ -50,17 +50,24 @@ public function setAbstractRectorDependencies(PropertyAdder $propertyAdder, Expr
         $newNode = $this->refactor($node);
+
+        // nothing has changed, return original node
         if ($newNode !== null) {
```
Wrong comment? If `$newNode` is different from null, it means we have a new Node

carusogabriel commented on src/Rector/AbstractRector.php:
```diff
@@ -50,17 +50,24 @@ public function setAbstractRectorDependencies(PropertyAdder $propertyAdder, Expr
+        if (! in_array($nodeClass, $this->getNodeTypes(), true)) {
+            return null;
         }
+//        if (method_exists($this, 'getNodeTypes')) {
```
TODO: Remove it and add `getNodeTypes` as an abstract method

TomasVotruba commented:
- duplicated old `instanceof` removed
- rector `MergeIsCandidateRector` moved to examples
- added `Rector\Utils\NodeTraverser\CallableNodeTraverser` that helps with inner traversing
- rebased

Ready to merge :+1:

carusogabriel commented:
@TomasVotruba I'll review this afternoon and I'll merge if everything is all right :+1:

TomasVotruba commented:
:+1: Thanks @carusogabriel
